tools/diagnostics/nsdiff

```diff
@@ -0,0 +1,875 @@
+              state 'signature'(s). Therefore, scrub state_signature LevelDB
+              entries and exclude 'participants' and 'signature' fields;
+              preserving only the 'value' JSON element.
+           3. Scrub the node_info; preserving only the 'view' field.
+           4. Extract the last/latest (view no, ppSeqNo) tuple from the logs
+              emitted by replica.py and node.py. TODO: update this description
+              once we have concrete examples.
+           5. All other files and directories are unaltered.
+        """
+        logger.debug("Transforming and scrubbing {} into {}".format(sin, sout))
+    
+        # The set of filenames that must exist in a directory to be assumed a
+        # LevelDB/RocksDB instance. Will be used in the following for-loop.
+        leveldb_rocksdb_filenames = [ 'CURRENT', 'LOG' ]
+
+        # Traverse directory sin
+        # TODO: 1. Optimize using multiprocessing module. At minimum, all
+        #       directories can be transformed and scrubbed in parallel.
+        #       https://docs.python.org/3/library/multiprocessing.html
+        #
+        #       2. Files in Non-LevelDB/RocksDB directories can be transformed
+        #          and scrubbed in parallel.
+        for dirName, subdirList, fileList in os.walk(abspath(sin)):
+            if (dirName in self.skipdirs):
+                logger.debug("Skipping dirName: {}".format(dirName))
+                break
+
+            if self.log_level <= logging.INFO:
+                logger.info("dirName: {}".format(dirName))
+
+            # Skip specific directories
+            subdirList[:] = [x for x in subdirList if x not in self.skipdirs]
+            # Skip specific files
+            fileList[:] = [x for x in fileList if x not in self.skipfiles]
+    
+            # Check if all of the leveldb_rocksdb_filenames are found in
+            # fileList set is_leveldb_rocksdb_dir and break if any of the
+            # filenames are not found
+            is_leveldb_rocksdb_dir = True
+            for filename in leveldb_rocksdb_filenames:
+                if filename not in fileList:
+                    is_leveldb_rocksdb_dir = False
+                    break
+    
+            (dirname, filename, name, extension) = self.split_name(dirName)
+            # Do not include recordings in diff
+            if (filename == "recorder" and dirname.endswith(
+                "/{}".format(sname))):
+                subdirList[:] = []
+                continue
+
+            if (is_leveldb_rocksdb_dir):
+                logger.debug("{} is a LevelDB/RocksDB directory!"
+                    .format(dirName))
+
+                # Use "nodename" in place of the node's name when a directory
+                # (dirName in this case) matches the node name (sname)
+                if (sname == filename):
+                    filename = "nodename"
+
+                if (filename.endswith("_transactions")):
```

Comment:
For me it looks like the method is overcomplicated. The type of the transaction log, hash store and other storages can be get from config file. Why don't we get it from there instead of trying all variants in try-else style?

Reply:
Initial we were not capturing the configuration. (not part of nsdiff original scope) So that is why we had a try-fail cycle. I think the try-fail gives us better resiliency at not much cost in the complexity of the code. But we can add TODOs to simplify this. But I don't think that this should hold this back.